hydra-node/test/Hydra/Port/SpecHook.hs

```diff
@@ -0,0 +1,11 @@
+module Hydra.Port.SpecHook where
+
+import Hydra.Prelude
+import Test.Hydra.Prelude
+
+import Control.Monad.Class.MonadSTM (newTVarIO)
+
+-- | Hspec hook to re-use the same 'TVar' for holding the list of all used ports.
+-- This is needed for the tests that depend on having a free port.
+hook :: SpecWith (TVar IO [Int]) -> Spec
+hook = beforeAll (newTVarIO [])
```

Comment:
Must: find a different way to re-use a port counting variable across tests

We cannot just force all tests to live in the same directory if they need a free port. What if there is another resource which needs to be managed to avoid test pollution?

One option (maybe the only one I know) is to use `unsafeDupablePerformIO` to give you a CAF of `TVar IO [Int]` in this case. Something like `globalUsedPorts = unsafeDupablePerformIO $ newTVarIO []`. See also https://hackage.haskell.org/package/global-variables-1.0.1.1/docs/Data-Global.html for a package which uses this approach. A well encapsulated, marked as "Test-only" (!) module for providing free ports should do the trick here.